libcodechecker/analyze/log_parser.py

```diff
@@ -210,18 +220,122 @@ def get_compiler_target(parseLogOptions, compiler):
     else:
         err = load_compiler_info(parseLogOptions.compiler_target_file,
-                                 compiler)
+                                 compiler,
+                                 'target')
 
     if parseLogOptions.output_path is not None:
-        dump_compiler_info(parseLogOptions.output_path,
-                           compiler_target_dump_file,
-                           {compiler: err})
+        dump_compiler_info(compiler_info_dump_file,
+                           compiler,
+                           'target',
+                           err)
     return parse_compiler_target(err)
 
 
-def remove_file_if_exists(filename):
-    if os.path.isfile(filename):
-        os.remove(filename)
+def get_compiler_standard(parseLogOptions, compiler, lang):
+    """
+    Returns the default compiler standard of the given compiler. The standard
+    is determined by the values of __STDC_VERSION__ and __cplusplus predefined
+    macros. These values are integers indicating the date of the standard.
+    However, GCC supports a GNU extension for each standard. For sake of
+    generality we return the GNU extended standard, since it should be a
+    superset of the non-extended one, thus applicable in a more general manner.
+    """
+    VERSION_C = u"""
+#include <stdio.h>
+
+int main()
+{
+  int version = 90;
+
+#if defined(__STDC_VERSION__)
+#  if __STDC_VERSION__ >= 201710L
+  version = 17;
+#  elif __STDC_VERSION__ >= 201112L
+  version = 11;
+#  elif __STDC_VERSION__ >= 199901L
+  version = 99;
+#  elif __STDC_VERSION__ >= 199409L
+  version = 95;
+#  endif
+#endif
+
+  printf("%d\\n", version);
+
+  return 0;
+}
+    """
+
+    VERSION_CPP = u"""
+#include <iostream>
+
+int main()
+{
+  int version = 98;
+
+#if defined(__cplusplus)
+#  if __cplusplus >= 201703L
+  version = 17;
+#  elif __cplusplus >= 201402L
+  version = 14;
+#  elif __cplusplus >= 201103L
+  version = 11;
+#  elif __cplusplus >= 199711L
+  version = 98;
+#  endif
+#endif
+
+  std::cout << version << std::endl;
+
+  return 0;
+}
+    """
+
+    standard = ""
+    if parseLogOptions.compiler_info_file is None:
+        with tempfile.NamedTemporaryFile(
+                suffix=('.c' if lang == 'c' else '.cpp')) as source:
+
+            with source.file as f:
+                f.write(VERSION_C if lang == 'c' else VERSION_CPP)
+
+            exe = os.path.join(tempfile.gettempdir(),
+                               next(tempfile._get_candidate_names()))
+
+            try:
+                proc = subprocess.Popen([compiler, '-o', exe, source.name],
+                                        stdout=subprocess.PIPE,
+                                        stderr=subprocess.PIPE)
+                proc.communicate()  # Wait for execution.
+            except OSError:
+                LOG.error("Error during the compilation of compiler standard "
+                          "detector.")
+
+        if os.path.isfile(exe):
+            os.chmod(exe, stat.S_IXUSR)
```

Comment:
Will the user always have the permission to set these bits?

Reply:
The idea of using preprocessing time error sounds good. Do you know any difference between `#error` and `#pragma GCC error`? Isn't the former one more standard?

Reply:
The `#pragma` version gives me different outputs with different compilers. `g++` (5.5.0) seems not to understand it: `error: invalid "#pragma GCC error" directive`.
By the way can we rely on the error message format between different compilers, or even same compiler and different versions? Wouldn't it complicate the parsing of this output?

Reply:
I didn't know about `#error`.

The output should be sufficiently small: `CC_FOUND_STARDARD_VER|23`. On that, a simple regex can give us the answer, irrespective to whatever specific compiler and version is used.